Bug fixes & new/expanded test cases

a-alamoudi opened the pull request:
Related to issues found in https://github.com/molinch/ConfigurationSubstitutor/issues/11#issuecomment-1003237216
- [FIX] Applied fixes to `EnableSubstitutionsWithDelimitedFallbackDefaults` to match behaviour of `EnableSubstitutions` when handling non-resolvable variables and `null` substituted values.
- [TEST] Added and expanded a couple of test cases to ensure the latest fixes.

a-alamoudi committed:
Bug fixes & new/expanded test cases

- [FIX] Applied fixes to `EnableSubstitutionsWithDelimitedFallbackDefaults` to match behaviour of `EnableSubstitutions` when handling non-resolvable variables and `null` substituted values.
- [TEST] Added and expanded a couple of test cases to ensure the latest fixes.

molinch commented on ConfigurationSubstitutor/ConfigurationSubstitutor.cs:
```diff
@@ -72,22 +72,21 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
                     {
-                        return capture.Value;
+                        value = value.Replace(capture.Value, delimitedVals[0]);
+                        // if declared EV isn't resolvable, assign it to provided fallback default value
```
Could you elaborate on what means EV?

a-alamoudi replied:
Absolutely, 'EV' refers to an environment variable. This section deals with the two tokens produced from breaking the captured value by the specified delimiter. Once execution arrives at this code segment, we're sure the delimiter matches, and that the first part (i.e., `delimitedVals[0]`) is the environment variable (EV).
Let's take the following example:
```
"Foo" = "${Var1:default_val}"
```
Substituting `config["Foo"]` with the start mark specified as `${`, the end mark as `}` and delimiter as `:`; we'd end up with `Var1` as the EV.

molinch replied:
Thanks for the explanation

molinch commented on ConfigurationSubstitutor/ConfigurationSubstitutor.cs:
```diff
@@ -72,22 +72,21 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
+                        // if declared EV isn't resolvable, assign it to provided fallback default value
+                        if (String.IsNullOrEmpty(configuration[delimitedVals[0]]))
+                        {
+                            configuration[delimitedVals[0]] = delimitedVals[1];
```
I know this was already added in former PR, but I somehow missed that.
Not a big fan of altering source configuration, if we can keep the function pure and pass the value downwards sounds simpler/better to reason about. What do you think?

a-alamoudi committed:
Avoid altering source configuration

- [REFACTOR] Avoid adding new items to source configuration and use local variables instead.

a-alamoudi commented:
https://github.com/molinch/ConfigurationSubstitutor/pull/13#discussion_r776949125 @molinch I took another crack at it and avoided altering the source configuration by using a local dictionary instead. It passes all of the existing tests and behaves just the same. Please review and let me know what you think.
Thanks, and Happy New Year! 🙂

Commit:
minor changes

molinch reviewed (approved):
Only did minor changes, thanks a lot @a-alamoudi !

molinch commented:
On nuget as 1.4.1, have a very nice 2022 year!

a-alamoudi commented:
@molinch Awesome! Thanks a lot, you too 😊
